### PR TITLE
Implemented issue #253 Feature request: query user to input login if …

### PR DESCRIPTION
**Pull request by deXol**
…field is empty

If there is a new credential in the browser without login name, then daemon sends a signal to GUI for request the login from the user. When the user added his/her login name in the MessageBox, then the GUI sends a message to the daemon with the given login name.

An example:
loginrequesttwitterexample

### CODE REVIEW

**Comment by raoulh** [comment]
Hi,

I don't really like this.

- First I do have some website where I have no login at all. Right now with this PR, I can't save any password for this website anymore. The only way is to put a random login. But then it enters the login when using the device alone. I mean when using the device alone with a credential that has no login, it only sends the password and do not ask for the login at all which is nice.

- Second, it will break and do nothing if you don't have any GUI running.

For me a "feature" like that should be implemented in the client which would send `the set_credential``at the beginning. Like here it's the extension that would need to ask the user if it's ok for him to use no login for this credential or query one directly.

**Comment by deXol** [comment]
Hello,

I suggested the extension solution too.
But if we stay with this solution, I can add a flag for set_credential to save without a login in this two case.
So if the GUI is not running or the user manually clicked OK on an empty login, I add this flag and process it with empty login as before.

**Comment by limpkin** [comment]
Makes sense to me!

As a reminder, this feature was requested because we want to migrate all
prompts to moolticute as some browsers do not allow prompts

File: src/MainWindow.cpp [hunk 1]
```diff
@@ -1038,6 +1039,36 @@ void MainWindow::wantExitFilesManagement()
...
+{
+    QJsonParseError err;
+    QJsonDocument jdoc = QJsonDocument::fromJson(message.toUtf8(), &err);
+    if (err.error != QJsonParseError::NoError)
+    {
+        qWarning() << "JSON parse error " << err.errorString();
+        return;
+    }
+    QJsonObject rootobj = jdoc.object();
+    QJsonObject o = rootobj["data"].toObject();
```
Comment (raoulh) [inline src/MainWindow.cpp]: All the json decoding should stay in WSClient. It's part of the json/websocket procotol, it should not be done in mainwindow

**Comment by deXol** [comment]
@raoulh Ok, I will fix these issues and then I create a new bigger dialog with those informations.

I am working on #255 which also requires very similar changes. Then I guess that also requires a new window or should I make this a generic one? (Texts are different there, and combobox instead of input field)

File: src/MainWindow.cpp (hunk 1 above)
Comment (deXol) [reply src/MainWindow.cpp]: moved to WSClient

**Comment by deXol** [comment]
Added empty logins, now working on the new dialog.

**Comment by raoulh** [comment]
@deXol #255 needs to be handled using the system notification from the OS. Not using our own dialogs. I will write some details about that on the issue

File: src/WSServer.cpp [hunk 2]
```diff
@@ -101,6 +102,20 @@ void WSServer::notifyClients(const QJsonObject &obj)
...
 }
 
+void WSServer::notifyGUI(const QString &message, bool &isGuiRunning)
+{
+    for (auto it = wsClients.begin(); it != wsClients.end(); ++it)
+    {
+        //Notify GUI
+        if (it.key()->resourceName().contains("localhost") && !(it.key()->origin().contains("extension")))
+        {
+            it.value()->sendJsonMessage(message);
```
Comment (raoulh) [inline src/WSServer.cpp]: `isGuiRunning` should be set to true here, it's bad to let it be initialized to a value from the caller of this function.

**Comment by limpkin** [comment]
Quick note for the future: imagine the following scenario:
- user visits a website for which he has credentials, but is only requested
his password (eg Google)
- the password submit is detected by the extension, which requests the user
for the login

For this pr as well as for the upcoming subdomain one, I therefore suggest
to not request the user anything if a previous credential request was
previously approved for the same complete URL.

File: src/WSServer.cpp (hunk 2 above)
Comment (deXol) [reply src/WSServer.cpp]: thanks, I forgot to add it there.

**Comment by deXol** [comment]
> Quick note for the future: imagine the following scenario: - user visits a website for which he has credentials, but is only requested his password (eg Google) - the password submit is detected by the extension, which requests the user for the login For this pr as well as for the upcoming subdomain one, I therefore suggest to not request the user anything if a previous credential request was previously approved for the same complete URL. On Mon, Sep 17, 2018, 13:10 mathieu.stephan@gmail.com < mathieu.stephan@gmail.com> wrote:
> […](#)

For a few days I have access only to a Mac, I tried google (gmail) with Safari, and if I try to login the daemon is not receiving the set_credential message. If I try to choose the login and password field manually with the extension, then I cannot do that neither, because only the site darken and cannot click anywhere.
I will try this on Windows in a few days.

**Comment by deXol** [comment]
Added the new login request dialog:
screen shot 2018-09-18 at 9 55 36 pm
Now we can add empty login name, it will accepted.
And if we choose cancel it will cancel the storage request.
If you have any other request regarding the dialog please let me know.

File: src/RequestLoginNameDialog.cpp [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    QDialog(parent),
+    ui(new Ui::RequestLoginNameDialog)
+{
+    ui->setupUi(this);
+    setFixedSize(size());
+    setWindowTitle(tr("Request Login Name"));
+    ui->loginLabel->setTextFormat(Qt::RichText);
+    ui->loginLabel->setText(tr("Login name for ") + "<b>" + service + "</b>:");
+    const QString informationText = "<h2><b>"
+                                  + tr("A credential without a login has been detected.")
```
Comment (raoulh) [inline src/RequestLoginNameDialog.cpp]: Change the text to:

`<h2><b>A credential without a login has been detected.</b></h2><br><br>You have the possibility to include a login before storing the credential in the device. If you don't need a login, you can ignore the request.`

Also, could you include an `ignore` button in the buttonbox

**Comment by deXol** [comment]
Added text and ignore button:
screen shot 2018-09-19 at 10 32 35 pm
If we click Ignore it is saving the empty login, and click on Cancel or escape is canceling the storage request.

**Comment by deXol** [comment]
> During testing I found an issue on mac.
> If we open any dialog from MainWindow, close it, hide MC and we are trying to show it again, then it will crash.
> It looks independent from this new dialog, because I could reproduce with Setting/Password Profiles... also every time I tried.
> It is crashing in AppGui.cpp:382
> `utils::mac::orderFrontRegardless(win->winId(), true);`

Fixed in the last commit.
According the callstack this problem caused the crash in the following ticket too: https://github.com/limpkin/mooltipass/issues/558

**Comment by raoulh** [comment]
@deXol Please remove the commit "Fixing crash on Mac from orderFrontRegardless" from this PR, and open an other PR for this. It does not belong here. thanks

**Comment by deXol** [comment]
> @deXol Please remove the commit "Fixing crash on Mac from orderFrontRegardless" from this PR, and open an other PR for this. It does not belong here. thanks

reverted it

**Comment by limpkin** [comment]
@raoulh any missing changes?

**Comment by deXol** [comment]
Added extension_version check in set_credential message from the browser extension according to the following comment: https://github.com/mooltipass/moolticute/issues/255#issuecomment-423836246 in #255

File: src/WSServerCon.cpp [hunk 4]
```diff
@@ -225,7 +225,7 @@ void WSServerCon::processMessage(const QString &message)
     {
         QJsonObject o = root["data"].toObject();
         QString loginName = o["login"].toString();
-        if (loginName.isEmpty() && !o.contains("saveConfirmed"))
+        if (loginName.isEmpty() && o.contains("extension_version") && !o.contains("saveConfirmed"))
```
Comment (raoulh) [inline src/WSServerCon.cpp]: I'm against this. First it does not belong to this PR, second if we use the `mc-cli` or the API to set an empty login we do not get the prompt.

Comment (limpkin) [reply src/WSServerCon.cpp]: first point: agreed, though we eventually need to integrate it somewhere
second point: I had forgotten about mc-cli. Could we add a "mc_cli_version" key and do a OR on that condition?

**Comment by raoulh** [comment]
Lets merge this now. Thanks @deXol
